pallets/transfer-allowlist/src/lib.rs

```diff
@@ -0,0 +1,481 @@
+			receiver: Location,
+		},
+		/// Event for Allowance delay update
+		TransferAllowanceDelaySet {
+			sender_account_id: T::AccountId,
+			currency_id: T::CurrencyId,
+			delay: T::BlockNumber,
+		},
+		/// Event for Allowance delay removal
+		TransferAllowanceDelayRemoval {
+			sender_account_id: T::AccountId,
+			currency_id: T::CurrencyId,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		/// Adds a transfer allowance for a sending Account/Currency.
+		/// Allowance either starts at the current block + the delay set for the account,
+		/// if a delay is present.
+		/// or block 0 if no delay is present.
+		/// Important! Account/Currency sets with an allowance set are restricted to just the allowances added for the account -
+		/// to have unrestricted transfers allowed for the sending Account and Currency, no allowances should be present.
+		///
+		/// Running this for an existing allowance generates a new allowance based on the current delay, or lack thereof
+		pub fn add_transfer_allowance(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			receiver: Location,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			let allowance_details = match Self::sender_currency_delay(&account_id, currency_id) {
+				Some(delay) => AllowanceDetails {
+					allowed_at: <frame_system::Pallet<T>>::block_number().saturating_add(delay),
+					..AllowanceDetails::default()
+				},
+				_ => AllowanceDetails::default(),
+			};
+			if !<AccountCurrencyTransferAllowance<T>>::contains_key((
+				&account_id,
+				&currency_id,
+				&receiver,
+			)) {
+				Self::increment_or_create_allowance_count(&account_id, &currency_id)?;
+				T::ReserveCurrency::reserve(&account_id, T::Deposit::get())?;
+			};
+			<AccountCurrencyTransferAllowance<T>>::insert(
+				(&account_id, &currency_id, &receiver),
+				&allowance_details,
+			);
+
+			Self::deposit_event(Event::TransferAllowanceCreated {
+				sender_account_id: account_id,
+				currency_id,
+				receiver,
+				allowed_at: allowance_details.allowed_at,
+				blocked_at: allowance_details.blocked_at,
+			});
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		/// Restricts a transfer allowance for a sending account/currency/receiver location to:
+		/// - either the current block + delay if a delay is set
+		/// - or the current block if no delay is set
+		pub fn remove_transfer_allowance(
+			origin: OriginFor<T>,
+			currency_id: T::CurrencyId,
+			receiver: Location,
+		) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			let blocked_at = match Self::sender_currency_delay(&account_id, currency_id) {
+				Some(delay) => <frame_system::Pallet<T>>::block_number().saturating_add(delay),
+				_ => <frame_system::Pallet<T>>::block_number(),
+			};
```

Comment:
I think this would be a bit easier to read with a `.map().unwrap_or()` pattern. Up to you 👍

Reply:
With the merging of the delay and count storage items, there's another option to check/destructure, so I've kept the match, but I agree with how it was before `.map().unwrap_or()` would be clearer. :)